token/cli/src/main.rs

```diff
@@ -380,62 +286,37 @@ fn command_thaw(config: &Config, account: Pubkey) -> CommandResult {
 
     println!("Freezing account: {}\n  Token: {}", account, token);
 
-    let mut transaction = Transaction::new_with_payer(
-        &[thaw_account(
-            &spl_token::id(),
-            &account,
-            &token,
-            &config.owner.pubkey(),
-            &[],
-        )?],
-        Some(&config.fee_payer.pubkey()),
-    );
-
-    let (recent_blockhash, fee_calculator) = config.rpc_client.get_recent_blockhash()?;
-    check_fee_payer_balance(
-        config,
-        fee_calculator.calculate_fee(&transaction.message(), None),
-    )?;
-    let mut signers = vec![config.fee_payer.as_ref(), config.owner.as_ref()];
-    unique_signers!(signers);
-    transaction.sign(&signers, recent_blockhash);
-    Ok(Some(transaction))
+    let instructions = vec![thaw_account(
+        &spl_token::id(),
+        &account,
+        &token,
+        &config.owner,
+        &[],
+    )?];
+    Ok(Some((0, instructions)))
 }
 
-fn command_wrap(config: &Config, sol: f64) -> CommandResult {
-    let account = Keypair::new();
+fn command_wrap(config: &Config, sol: f64, account: Pubkey) -> CommandResult {
     let lamports = sol_to_lamports(sol);
-    println!("Wrapping {} SOL into {}", sol, account.pubkey());
-
-    let mut transaction = Transaction::new_with_payer(
-        &[
-            system_instruction::create_account(
-                &config.owner.pubkey(),
-                &account.pubkey(),
-                lamports,
-                Account::LEN as u64,
-                &spl_token::id(),
-            ),
-            initialize_account(
-                &spl_token::id(),
-                &account.pubkey(),
-                &native_mint::id(),
-                &config.owner.pubkey(),
-            )?,
-        ],
-        Some(&config.fee_payer.pubkey()),
-    );
+    println!("Wrapping {} SOL into {}", sol, account);
 
-    let (recent_blockhash, fee_calculator) = config.rpc_client.get_recent_blockhash()?;
+    let instructions = vec![
+        system_instruction::create_account(
+            &config.owner,
+            &account,
+            lamports,
+            Account::LEN as u64,
+            &spl_token::id(),
+        ),
+        initialize_account(
+            &spl_token::id(),
+            &account,
+            &native_mint::id(),
+            &config.owner,
+        )?,
+    ];
     check_owner_balance(config, lamports)?;
```

Comment:
This has been bothering me, because it may not be correct if fee_payer == owner (owner balance may be enough to cover lamports and fee separately, but not both). It's a little bit less obvious with the fee_payer check moved. Got any clever ideas to fix?

Reply:
I've been thinking of adding `owner_lamports` as a third field (Likely to be promoted to a struct) of the `CommandResult` success tuple and doing the check down in the `.and_then()`, where there will be sign-only gating soon